client/async.go

```diff
@@ -0,0 +1,268 @@
+				hostname,
+				method,
+				e.Topic,
+			).Observe(timeUsed)
+		}
+		l.WithFields(logrus.Fields{
+			"timeUsed": timeUsed,
+			"reply":    reply.(*pb.SendEventsResponse),
+		}).Debug("request processed")
+	}(time.Now())
+
+	err := invoker(ctx, method, req, reply, cc, opts...)
+	if err != nil {
+		l.WithError(err).Error("error processing request")
+		for _, e := range events {
+			metrics.ClientRequestsFailureCounter.WithLabelValues(
+				hostname,
+				method,
+				e.Topic,
+				err.Error(),
+			).Inc()
+		}
+	} else {
+		for _, e := range events {
+			metrics.ClientRequestsSuccessCounter.WithLabelValues(
+				hostname,
+				method,
+				e.Topic,
+			).Inc()
+		}
+	}
+	return err
+}
+
+func (a *gRPCClientAsync) send(ctx context.Context, event *pb.Event) error {
+	atomic.AddUint64(&a.transientCounter, 1)
+	a.eventsChannel <- event
+	return nil
+}
+
+func (a *gRPCClientAsync) sendRoutine() {
+	ticker := time.NewTicker(a.flushInterval)
+	defer ticker.Stop()
+
+	req := &pb.SendEventsRequest{}
+	req.Events = make([]*pb.Event, 0, a.flushSize)
+
+	send := func() {
+		cpy := req
+		cpy.Id = uuid.NewV4().String()
+		req = &pb.SendEventsRequest{}
+		req.Events = make([]*pb.Event, 0, a.flushSize)
+		go a.sendEvents(cpy, 0)
+	}
+
+	for {
+		select {
+		case e := <-a.eventsChannel:
+			if len(req.Events) == 0 {
+				a.wg.Add(1)
+			}
+			atomic.AddUint64(&a.transientCounter, ^uint64(0))
+			req.Events = append(req.Events, e)
+			if len(req.Events) == a.flushSize {
+				send()
+			}
+		case <-ticker.C:
+			if len(req.Events) > 0 {
+				send()
+			}
+		}
+	}
+}
+
+func (a *gRPCClientAsync) sendEvents(req *pb.SendEventsRequest, retryCount int) {
+	l := a.logger.WithFields(logrus.Fields{
+		"operation":  "sendEvents",
+		"requestId":  req.Id,
+		"retryCount": retryCount,
+		"size":       len(req.Events),
+	})
+	if retryCount > a.maxRetries {
+		l.Info("dropped events due to max retries")
+		a.wg.Done()
+		return
+	}
+	ctx, _ := context.WithTimeout(context.Background(), a.timeout)
+	// in case server's producer fail to send any event, failure indexes are sent
+	// in response to be retried
+	res, err := a.client.SendEvents(ctx, req)
+	if err != nil {
+		l.WithError(err).Error("failed to send events")
+		time.Sleep(time.Duration(math.Pow(2, float64(retryCount))) * a.retryInterval)
+		a.sendEvents(req, retryCount+1)
+	}
+	if res.FailureIndexes != nil && len(res.FailureIndexes) != 0 {
+		l.WithFields(logrus.Fields{
+			"failureIndexes": res.FailureIndexes,
+		}).Error("failed to send events")
+		time.Sleep(time.Duration(math.Pow(2, float64(retryCount))) * a.retryInterval)
+		events := make([]*pb.Event, 0, len(res.FailureIndexes))
+		for _, index := range res.FailureIndexes {
+			events = append(events, req.Events[index])
+		}
+		req.Events = events
+		a.sendEvents(req, retryCount+1)
+	} else {
+		a.wg.Done()
+	}
+}
+
+// Wait on pending async send of events
+func (a *gRPCClientAsync) Wait() {
+	for {
+		if atomic.LoadUint64(&a.transientCounter) == 0 {
```

Comment:
Could we do better than busy waiting? A WaitGroup seems better suited to counting remaining tasks.